[mlir][linalg] `LinalgOp`: Disallow mixed tensor/buffer semantics

matthias-springer opened the pull request:
Related discussion: https://github.com/llvm/llvm-project/pull/73908/files#r1414913030.

This change fixes #73547.

jpienaar reviewed (approved):
This looks good given discussion thread/Nicholas's expectations here.

matthias-springer committed:
[mlir][linalg] `LinalgOp`: Disallow mixed tensor/buffer semantics

Related discussion: https://github.com/llvm/llvm-project/pull/73908/files#r1414913030.

This change fixes #73547.
